Fix/lectern

ShaneBeee opened the pull request:
- Allow all players to read books in lecterns
- Prevent players without container trust from removing books from/placing books into lectern

References #537 

**NOTE**:
Don't ask me why, maybe a Spigot/Bukkit mistake, but when trying to put a book into a lectern, it calls the BlockPlaceEvent (weird hey?!?)

ShaneBeee committed:
Lectern Books - changes
- Allow all players to read books in lecterns
- Prevent players without container trust from removing books from/placing books into lectern

ShaneBeee committed:
Fix wildcard imports

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java:
```diff
@@ -232,6 +232,24 @@ public void onBlockPlace(BlockPlaceEvent placeEvent)
+			Material offHand = player.getInventory().getItemInOffHand().getType();
+			PlayerData playerData = this.dataStore.getPlayerData(player.getUniqueId());
+			Claim claim = this.dataStore.getClaimAt(block.getLocation(), true, playerData.lastClaim);
+			if (block.getType() == Material.LECTERN && mainHand != Material.LECTERN && offHand != Material.LECTERN)
```
Shouldn't the hand material be checked against `BlockPlaceEvent#getHand`? This check could fail when holding a lectern in one hand and a book in the other. Weird edge case, but technically could block valid book insertion.
Could check that `BlockPlaceEvent#getReplacedState`'s type is also a lectern instead of checking hands.

ShaneBeee replied:
Good call, Ill do some testing and make the changes.

ShaneBeee replied:
Well, that was a surprisingly easy fix.

ShaneBeee committed:
Lectern - changes
- Changes suggested by Jikoo

RoboMWM commented on src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java:
```diff
@@ -232,6 +233,22 @@ public void onBlockPlace(BlockPlaceEvent placeEvent)
 		String noBuildReason = GriefPrevention.instance.allowBuild(player, block.getLocation(), block.getType());
 		if(noBuildReason != null)
 		{
+			// Prevent players without container trust from putting books into lecterns
```
`// Allow players with container trust to place books in lecterns`

RoboMWM commented on src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:
```diff
@@ -2620,6 +2621,21 @@ else if(playerData.shovelMode == ShovelMode.Subdivide)
 	}
+
+	// Stops an untrusted player from removing a book from a lectern
+	@EventHandler
```
`(priority = EventPriority.LOWEST, ignorecancelled = true)`

ShaneBeee committed:
Lectern - changes
- Changes requested by Robo

ShaneBeee committed:
Lectern - Options
- New config option, to require access trust to read lectern books, true by default

ShaneBeee committed:
Lectern - note change
- Changed the note as requested by Robo

RoboMWM commented on src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:
```diff
@@ -135,6 +135,8 @@
 	public boolean config_claims_firespreads;						//whether fire will spread in claims
 	public boolean config_claims_firedamages;						//whether fire will damage in claims
+
+	public boolean config_claims_lecternreading;					//reading lecterns requires access trust
```
To better match a similar config for villagers, `config_claims_lecternReadingRequiresAccessTrust` (i.e. make it similar to config node key)

ShaneBeee replied:
what? you don't like my super long field name? (jk jk) 
Yeah, ill change it up right quick 👍 

scratch that I read that backwards haha. I thought mine was the long one

RoboMWM commented on src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:
```diff
@@ -2620,6 +2636,21 @@ else if(playerData.shovelMode == ShovelMode.Subdivide)
+		Player player = event.getPlayer();
+		PlayerData playerData = this.dataStore.getPlayerData(player.getUniqueId());
+		Claim claim = this.dataStore.getClaimAt(event.getLectern().getLocation(), false, playerData.lastClaim);
+		if (claim.allowContainers(player) != null)
```
You need to nullcheck here.

RoboMWM replied:
And if it isn't null, then you can 
- set player#lastClaim
- check allowContainers

ShaneBeee replied:
Oh duh, how did I forget that. I guess I didn't test outside of a claim to see what I would assume would be a nasty error.

Ill push my fixes in a minute after I test to make sure they work.

ShaneBeee committed:
Lectern - field changed
- changed field name as requested by Robo

ShaneBeee committed:
Lectern - changes
- Add null check for claim itself

zedwick commented on src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java:
```diff
@@ -232,6 +233,22 @@ public void onBlockPlace(BlockPlaceEvent placeEvent)
+			Claim claim = this.dataStore.getClaimAt(block.getLocation(), true, playerData.lastClaim);
+			if (block.getType() == Material.LECTERN && placeEvent.getBlockReplacedState() instanceof Lectern)
+			{
+				if (claim.allowContainers(player) != null)
```
Should check if claim is not null here, and can set PlayerData#lastClaim if so.

ShaneBeee replied:
Good call, I dont know why I keep forgetting to check for claim being null.

zedwick commented on src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java:
```diff
@@ -232,6 +233,22 @@ public void onBlockPlace(BlockPlaceEvent placeEvent)
+				if (claim.allowContainers(player) != null)
+				{
+					placeEvent.setCancelled(true);
+					GriefPrevention.sendMessage(player, TextMode.Err, claim.allowContainers(player));
```
Perhaps minor, but you could store the result of `claim.allowContainers(player)` above and reuse the result, rather than running that check twice.

ShaneBeee replied:
You're right definitely better.

zedwick commented on src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:
```diff
@@ -2620,6 +2636,25 @@ else if(playerData.shovelMode == ShovelMode.Subdivide)
+			{
+				event.setCancelled(true);
+				player.closeInventory();
+				GriefPrevention.sendMessage(player, TextMode.Err, claim.allowContainers(player));
```
Store result of `claim.allowContainers(player)` and reuse, rather than running the check twice.

zedwick replied:
You didn't change this bit.

ShaneBeee replied:
WTF is wrong with me today... jeez haha! Clearly I need more coffee. Ill change it right up.

ShaneBeee committed:
Lectern - changes
- Changes as requested by zedwick

ShaneBeee committed:
Lectern - changes
- Changes as requested by zedwick

RoboMWM commented on src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java:
```diff
@@ -1658,7 +1659,22 @@ void onPlayerInteract(PlayerInteractEvent event)
+					{
+						return;
+					}
+				}
```
I actually prefer the change @zedwick made in #580 - he exempted reading from lecterns from the containertrust check, and added a case to check for them in the accesstrust check. Better than just a special section of code for a single case.

ShaneBeee replied:
You're right, that'll definitely clean up the code quite a bit

RoboMWM commented on src/main/java/me/ryanhamshire/GriefPrevention/BlockEventHandler.java:
```diff
@@ -232,6 +233,27 @@ public void onBlockPlace(BlockPlaceEvent placeEvent)
+					else
+					{
+						return;
+					}
```
would suggest reducing code via

```java
if (noContainerReason == null)
    return;
// else logic, but now no longer needs `else`
```

ShaneBeee replied:
Good call, that will clean that up quite a bit.

ShaneBeee committed:
Lectern - changes
- Changes as requested by Robo

ShaneBeee committed:
Lectern - small fix
- Fixed an issue with using clickedBlock.getType() throwing null

RoboMWM reviewed (approved):
Looks good to me at a glance.

RoboMWM committed:
Merge branch 'master' into fix/lectern
